src/ios/MultipeerConnectivity/THEMultipeerSocketRelay.m

```diff
@@ -0,0 +1,254 @@
+  [_inputStream scheduleInRunLoop:[NSRunLoop mainRunLoop] forMode:NSDefaultRunLoopMode];
+  [_inputStream open];
+  
+  _outputStream.delegate = self;
+  [_outputStream scheduleInRunLoop:[NSRunLoop mainRunLoop] forMode:NSDefaultRunLoopMode];
+  [_outputStream open];
+}
+
+- (BOOL)canCreateSocket
+{
+  // Postpone socket creation until we know we have somewhere to send
+  // it's data
+  return (_inputStream && _outputStream);
+}
+
+- (BOOL)tryCreateSocket
+{
+  // Base class only. Shouldn't be reachable
+  assert(NO);
+  return NO;
+}
+
+- (void)didCreateSocket:(GCDAsyncSocket *)socket
+{
+  // Socket's been created which means we can open up the stream
+  assert(_socket == nil);
+
+  _socket = socket;
+  [self openStreams];
+  [_socket readDataWithTimeout:-1 tag:0];
+}
+
+- (void)stop
+{
+  NSLog(@"%@ relay: stopping", _relayType);
+
+  if (_socket)
+  {
+    NSLog(@"%@ relay: destroying socket %p", _relayType, _socket);
+    _socket.delegate = nil;
+    [_socket disconnect];
+    _socket = nil;
+  }
+
+  if (_inputStream)
+  {
+    [_inputStream close];
+    [_inputStream removeFromRunLoop: [NSRunLoop currentRunLoop] forMode: NSDefaultRunLoopMode];
+    _inputStream = nil;
+  }
+
+  if (_outputStream)
+  {
+    [_outputStream close];
+    [_outputStream removeFromRunLoop: [NSRunLoop currentRunLoop] forMode: NSDefaultRunLoopMode];
+    _outputStream = nil;
+  }
+}
+
+- (void)dealloc
+{
+  NSLog(@"%@ relay: dealloc", _relayType);
+  [self stop];  
+}
+
+- (void)socket:(GCDAsyncSocket *)sock didReadData:(NSData *)data withTag:(long)tag
+{
+  assert(sock == _socket);
+  assert(_outputStream != nil);
+
+  if ([_outputStream write:data.bytes maxLength:data.length] != data.length)
+  {
+    NSLog(@"ERROR: Writing to output stream");
+  }
+
+  [_socket readDataWithTimeout:-1 tag:tag];
+}
+
+- (void)socket:(GCDAsyncSocket *)sock didWriteDataWithTag:(long)tag
+{
+}
+
+- (void)socketDidDisconnect:(GCDAsyncSocket *)sock withError:(NSError *)err
+{
+  assert(_socket == nil || sock == _socket);
+
+  // Usually benign, the upper layer just closed their connection
+  // they may want to connect again later
+
+  NSLog(@"%@ relay: socket disconnected", _relayType);
+
+  if (err) 
+  {
+      NSLog(@"%@ relay: %p disconnected with error %@ ", _relayType, sock, [err description]);
+  }
+}
+
+#pragma mark - NSStreamDelegate
+
+- (void)stream:(NSStream *)aStream handleEvent:(NSStreamEvent)eventCode
+{
+  if (aStream == _inputStream) 
+  {
+    switch (eventCode) 
+    {
+      case NSStreamEventOpenCompleted:
+      {
+        //NSLog(@"%@ relay: inputStream opened", _relayType);
+      }
+      break;
+
+      case NSStreamEventHasSpaceAvailable:
+      {
+        //NSLog(@"%@ relay: inputStream hasSpace", _relayType);
+      }
+      break;
+
+      case NSStreamEventHasBytesAvailable:
+      {
+        const uint bufferValue = 1024;
+        uint8_t *buffer = malloc(bufferValue);
+        NSInteger len = [_inputStream read:buffer maxLength:sizeof(bufferValue)];
```

Comment:
Are we guaranteed that the amount of data that will be available is never larger than 1K? In either case the magic number should be in a const and some comment about the guarantee would seem appropriate. If there is no guarantee than what happens if there is more than 1K available?